### ISSUE TITLE
Rule proposal: no-relative-import-of-aliased-path

### ISSUE THREAD

**Issue by JasonKleban** [typescript-eslint/typescript-eslint#834]
To support Project-References-based monorepos, disallow relative paths where tsconfig is configured to alias that path.

Example based on https://github.com/RyanCavanaugh/project-references-demo :

If path aliases are added:

```
"paths": { // in tsconfig-base.json
    "@peer-core/*": ["core/src/*"],
    "@peer-animals/*": ["animals/src/*"]
}
```

Then import statements of relative paths to those aliased locations should be an error:

```
import { Dog, createDog } from '../animals/index'; // in zoo/zoo.ts
```

Instead, the line should be fixed to:

```
import { Dog, createDog } from '@peer-animals'; // in zoo/zoo.ts
```

While **requiring** relative paths among files in the same project:

```
import Animal from './'; // in animals/dog.ts
```

but not: 

```
import Animal from '@peer-animals'; // in animals/dog.ts
```

**Comment by JasonKleban** [comment]
[RyanCavanaugh](https://github.com/RyanCavanaugh/project-references-demo )'s original with path aliases, `eslint` and `@typescript-eslint` added: https://github.com/JasonKleban/project-references-demo

**Comment by bradzacher** [comment]
I'm not sure, but would this be better suited to be built as part of `eslint-plugin-import`?

**Comment by JasonKleban** [comment]
Because of path aliases (and in support of project references), it does still seem to be typescript-specific to me.  🤷‍♂

**Comment by JoshuaKGoldberg** [comment]
There should probably be some thought put into when the rule should prefer a relative import vs. a path-based one. Given a choice between the following:

```ts
// ~/abc/def/ghi/file.ts
import sibling from './sibling';
import sibling from '~/abc/def/ghi/sibling';
```

...I think we'd probably prefer the relative import. 

Maybe a good starting default heuristic would be to only care about `..` imports?

_(not sure if I parsed the OP correctly, is that what you meant?)_

**Comment by JasonKleban** [comment]
@JoshuaKGoldberg are root-relative tildes standard?  Either way I hadn't considered discerning among all three but as long as it isn't an auto-fix we just have to disallow the relative vs package syntaxes.

**Comment by JoshuaKGoldberg** [comment]
I don't think they're standard, but I strongly prefer them 😄. Imports without any prefix or with the `@` prefix can be mistaken for existing packages:

```ts
// Is this from src/my/code? Or the 'my' package?
import stuff from 'my/code';
```

```ts
// Is this from src/my/code? Or the '@my/code' package?
import stuff from '@my/code';
```

Regardless, whatever the user puts in their `tsconfig.json` is what this rule would read from.

**Comment by JoshuaKGoldberg** [comment]
👋 Coming back to this after 2.5 years: we took another look and don't think this is a rule that belongs in typescript-eslint right now. There's a lot of potential for nuance that could boil down to user preference.

We're going to say that if this exists, it should be in a plugin more targeted towards imports. At the moment that's look like [`eslint-plugin-import`](https://github.com/import-js/eslint-plugin-import) or [`eslint-plugin-import-x`](https://github.com/un-ts/eslint-plugin-import-x).

Alternately, https://github.com/Adjective-Object/good-fences-rs-core is a nice project for establishing code boundaries.

Thanks all!
